File: python/open_guardrail/guards/profanity_fi.py
"""Detect Finnish profanity."""
from __future__ import annotations

import re
import time
from typing import Optional

from open_guardrail.core import GuardResult

_PATTERNS: list[re.Pattern[str]] = [
    re.compile(r"(?:^|\s)perkele(?:\s|$|[.,!?])", re.IGNORECASE),
    re.compile(r"(?:^|\s)vittu(?:\s|$|[.,!?])", re.IGNORECASE),
    re.compile(r"(?:^|\s)saatana(?:\s|$|[.,!?])", re.IGNORECASE),
    re.compile(r"(?:^|\s)paska(?:\s|$|[.,!?])", re.IGNORECASE),
    re.compile(r"(?:^|\s)helvetti(?:\s|$|[.,!?])", re.IGNORECASE),
    re.compile(r"(?:^|\s)kusipää(?:\s|$|[.,!?])", re.IGNORECASE),
    re.compile(r"(?:^|\s)mulkku(?:\s|$|[.,!?])", re.IGNORECASE),
    re.compile(r"(?:^|\s)huora(?:\s|$|[.,!?])", re.IGNORECASE),
    re.compile(r"(?:^|\s)kyrpä(?:\s|$|[.,!?])", re.IGNORECASE),
    re.compile(r"(?:^|\s)runkkari(?:\s|$|[.,!?])", re.IGNORECASE),
]


class _ProfanityFi:
    def __init__(self, *, action: str = "block") -> None:
        self.name = "profanity-fi"
        self.action = action

    def check(self, text: str, stage: str = "input") -> GuardResult:
        start = time.perf_counter()
        matched: list[str] = []
        for pat in _PATTERNS:
            if pat.search(text):
                matched.append(pat.pattern)
        triggered = len(matched) > 0
        score = min(len(matched) / 3, 1.0) if triggered else 0.0
        elapsed = (time.perf_counter() - start) * 1000
        return GuardResult(
            guard_name="profanity-fi",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=score,
            message="Finnish profanity detected" if triggered else None,
            latency_ms=round(elapsed, 2),
            details=(
                {"matched_patterns": len(matched)}
                if triggered
                else None
            ),
        )


def profanity_fi(*, action: str = "block") -> _ProfanityFi:
    return _ProfanityFi(action=action)

File: packages/guards/src/profanity-fi.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface ProfanityFiOptions {
  action: 'block' | 'warn';
}

const PROFANITY_PATTERNS: RegExp[] = [
  /(?:^|\s)perkele(?:\s|$|[.,!?])/i,
  /(?:^|\s)vittu(?:\s|$|[.,!?])/i,
  /(?:^|\s)saatana(?:\s|$|[.,!?])/i,
  /(?:^|\s)paska(?:\s|$|[.,!?])/i,
  /(?:^|\s)helvetti(?:\s|$|[.,!?])/i,
  /(?:^|\s)kusipää(?:\s|$|[.,!?])/i,
  /(?:^|\s)mulkku(?:\s|$|[.,!?])/i,
  /(?:^|\s)huora(?:\s|$|[.,!?])/i,
  /(?:^|\s)kyrpä(?:\s|$|[.,!?])/i,
  /(?:^|\s)runkkari(?:\s|$|[.,!?])/i,
];

export function profanityFi(options: ProfanityFiOptions): Guard {
  return {
    name: 'profanity-fi',
    version: '0.1.0',
    description: 'Detects Finnish profanity',
    category: 'locale',
    supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const matched: string[] = [];

      for (const pattern of PROFANITY_PATTERNS) {
        if (pattern.test(text)) {
          matched.push(pattern.source);
        }
      }

      const triggered = matched.length > 0;
      const score = triggered ? Math.min(matched.length / 3, 1.0) : 0;

      return {
        guardName: 'profanity-fi',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { matchedPatterns: matched.length } : undefined,
      };
    },
  };
}
